src/dmd/expressionsem.d

```diff
@@ -8650,9 +8650,62 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         }
         else
         {
+            // Try alias this on first operand
+            static Expression tryAliasThisForExp1(BinAssignExp exp, Scope* sc)
+            {
+                AggregateDeclaration ad1 = isAggregate(exp.e1.type);
+                if (ad1 && ad1.aliasthis)
+                {
+                    /* Rewrite (e1 op e2) as:
+                     *      (e1.aliasthis op e2)
+                     */
+                    if (exp.att1 && exp.e1.type == exp.att1)
+                        return null;
+                    //printf("att %s e1 = %s\n", Token::toChars(e.op), e.e1.type.toChars());
+                    Expression e1 = new DotIdExp(exp.loc, exp.e1, ad1.aliasthis.ident);
+                    BinExp be = cast(BinExp)exp.copy();
+                    if (!be.att1 && exp.e1.type.checkAliasThisRec())
+                        be.att1 = exp.e1.type;
+                    be.e1 = e1;
+                    return be.trySemantic(sc);
+                }
+                return null;
+            }
+
+            // Try alias this on second operand
+            static Expression tryAliasThisForExp2(BinAssignExp exp, Scope* sc)
+            {
+                AggregateDeclaration ad2 = isAggregate(exp.e2.type);
+                if (ad2 && ad2.aliasthis)
+                {
+                    /* Rewrite (e1 op e2) as:
+                     *      (e1 op e2.aliasthis)
+                     */
+                    if (exp.att2 && exp.e2.type == exp.att2)
+                        return null;
+                    //printf("att %s e2 = %s\n", Token::toChars(e.op), e.e2.type.toChars());
+                    Expression e2 = new DotIdExp(exp.loc, exp.e2, ad2.aliasthis.ident);
+                    BinExp be = cast(BinExp)exp.copy();
+                    if (!be.att2 && exp.e2.type.checkAliasThisRec())
+                        be.att2 = exp.e2.type;
+                    be.e2 = e2;
+                    return be.trySemantic(sc);
+                }
+                return null;
+            }
```

Comment:
This code is complex enough to warrant an attempt to reuse - you may want to pass e1/e2 and att1/att2 as extra parameters to a common private function.

Reply:
Per Kate Gregory's CppCon talk: instead of:

```
if (expr)
{
    ... interesting part ...
   return interesting;
}
return null;
```

prefer this:

```
if (!expr)
    return null;
... interesting part ...
return interesting;
```

Reply:
>This code is complex enough to warrant an attempt to reuse - you may want to pass e1/e2 and att1/att2 as extra parameters to a common private function.

That's not possible without an additional flag. Look at lines 8689 to 8691: the accesses to `be.att2` and `be.e2`  are done through a local variable. The only two solutions that come to mind are:

-> use a bool flag to choose the fields;
-> move the copy of exp outside of the function and pass be.att and be.e as parameters, but this will impose an overhead since the copy will be done for all expressions, not only for those that actually have an alias this defined.

I think that having 2 separate functions is a better alternative than employing elucubrated tricks to have a single function.